service/config_provider.go

```diff
@@ -68,50 +68,110 @@ type ConfigProvider interface {
+
+// Deprecated: [v0.49.0] use NewConfigProvider instead.
 func MustNewConfigProvider(
 	locations []string,
 	configMapProviders map[string]config.MapProvider,
 	cfgMapConverters []config.MapConverterFunc,
 	configUnmarshaler configunmarshaler.ConfigUnmarshaler) ConfigProvider {
+	opts := []ConfigProviderOption{
+		WithConfigMapConverters(cfgMapConverters),
+		WithConfigUnmarshaler(configUnmarshaler),
+	}
+	for _, c := range configMapProviders {
+		opts = append(opts, WithConfigMapProvider(c))
+	}
+	cfgProvider, err := NewConfigProvider(locations, opts...)
+	if err != nil {
+		panic(err)
+	}
+	return cfgProvider
+}
+
+// NewConfigProvider returns a new ConfigProvider that provides the configuration:
+// * Retrieve the config.Map by merging all retrieved maps from the given `locations` in order.
+// * Then applies all the config.MapConverterFunc in the given order.
+// * Then unmarshals the config.Map into the service Config.
+func NewConfigProvider(locations []string, opts ...ConfigProviderOption) (ConfigProvider, error) {
+	if len(locations) == 0 {
+		return nil, fmt.Errorf("cannot create ConfigProvider: no locations provided")
+	}
 	// Safe copy, ensures the slice cannot be changed from the caller.
 	locationsCopy := make([]string, len(locations))
 	copy(locationsCopy, locations)
-	return &configProvider{
+
+	opts = append([]ConfigProviderOption{
+		WithConfigMapProvider(filemapprovider.New()),
+		WithConfigMapProvider(envmapprovider.New()),
+		WithConfigMapProvider(yamlmapprovider.New()),
+	}, opts...)
+
+	provider := configProvider{
 		locations:          locationsCopy,
-		configMapProviders: configMapProviders,
-		cfgMapConverters:   cfgMapConverters,
-		configUnmarshaler:  configUnmarshaler,
-		watcher:            make(chan error, 1),
+		configMapProviders: make(map[string]config.MapProvider),
+		configMapConverters: []config.MapConverterFunc{
+			configmapprovider.NewExpandConverter(),
+		},
+		configUnmarshaler: configunmarshaler.NewDefault(),
+		watcher:           make(chan error, 1),
+	}
+
+	// Override default values with user options
+	for _, o := range opts {
+		o(&provider)
 	}
+
+	return &provider, nil
 }
 
-// MustNewDefaultConfigProvider returns the default ConfigProvider from slice of location strings
-// (e.g. file:/path/to/config.yaml) and property overrides (e.g. service.telemetry.metrics.address=localhost:8888).
+// Deprecated: [v0.49.0] use NewConfigProvider instead.
 func MustNewDefaultConfigProvider(configLocations []string, properties []string) ConfigProvider {
-	return MustNewConfigProvider(
+	cfgProvider, err := NewConfigProvider(
 		configLocations,
-		makeConfigMapProviderMap(filemapprovider.New(), envmapprovider.New(), yamlmapprovider.New()),
-		[]config.MapConverterFunc{
+		WithConfigMapConverters([]config.MapConverterFunc{
 			configmapprovider.NewOverwritePropertiesConverter(properties),
 			configmapprovider.NewExpandConverter(),
-		},
-		configunmarshaler.NewDefault())
+		}))
+	if err != nil {
+		panic(err)
```

Comment:
I guess this is so early in the code that we don't have even a basic logging facility set, right?

Reply:
Yes